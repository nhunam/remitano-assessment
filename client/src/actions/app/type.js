import keyMirror from "keymirror";

export const ActionTypes = keyMirror({
  CLEAR_MESSAGE: undefined
});
