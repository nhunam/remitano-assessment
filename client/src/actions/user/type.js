import keyMirror from "keymirror";

export const ActionTypes = keyMirror({
  USER_CREATE: undefined,
  USER_LOGIN: undefined,
  USER_LOGIN_SUCCESS: undefined,
  USER_LOGIN_FAILURE: undefined,
  USER_LOGOUT: undefined,
  USER_LOGOUT_SUCCESS: undefined,
  USER_LOGOUT_FAILURE: undefined
});
