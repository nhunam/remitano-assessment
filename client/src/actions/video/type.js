import keyMirror from "keymirror";

export const ActionTypes = keyMirror({
  VIDEO_FETCH: undefined,
  VIDEO_FETCH_SUCCESS: undefined,
  VIDEO_FETCH_FAILURE: undefined,
  VIDEO_SHARING: undefined,
  VIDEO_SHARING_SUCCESS: undefined,
  VIDEO_SHARING_FAILURE: undefined
});
