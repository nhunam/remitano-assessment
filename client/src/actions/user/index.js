import { ActionTypes } from "./type";
import { createActions } from "redux-actions";

export const {
  userLogin: login,
  userLogout: logout,
  userCreate: create
} = createActions({
  [ActionTypes.USER_LOGIN]: param => ({ ...param }),
  [ActionTypes.USER_CREATE]: param => ({ ...param }),
  [ActionTypes.USER_LOGOUT]: () => ({})
});
