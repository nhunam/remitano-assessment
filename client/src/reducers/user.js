import { ActionTypes } from "actions/user/type";
import { AppStatus } from "configurations";
import { handleActions } from "redux-actions";
import immutable from "immutability-helper";

export const userState = {
  isAuthenticated: false,
  info: {
    id: -1,
    email: "",
    token: ""
  }
};

export default {
  user: handleActions(
    {
      [ActionTypes.USER_LOGIN_SUCCESS]: (state, { payload }) =>
        immutable(state, {
          isAuthenticated: { $set: true },
          info: {
            id: { $set: payload.id },
            email: { $set: payload.email },
            token: { $set: payload.token }
          }
        }),
      [ActionTypes.USER_LOGOUT]: state =>
        immutable(state, {
          status: { $set: AppStatus.RUNNING }
        }),
      [ActionTypes.USER_LOGOUT_SUCCESS]: state =>
        immutable(state, {
          isAuthenticated: { $set: false },
          info: { id: { $set: -1 }, email: { $set: "" }, token: { $set: "" } },
          status: { $set: AppStatus.IDLE }
        })
    },
    userState
  )
};
