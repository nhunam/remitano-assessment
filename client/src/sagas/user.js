import { all, call, put, takeLatest } from "redux-saga/effects";

import { ActionTypes } from "actions/user/type";
import jwt_decode from "jwt-decode";
import { request } from "utils/client";

/**
 * Login
 */
export function* authenticate(param) {
  try {
    const payload = { method: "POST", ...param };
    const response = yield call(request, "/token", payload);
    const { token } = response;
    const user = jwt_decode(token);

    yield put({
      type: ActionTypes.USER_LOGIN_SUCCESS,
      payload: { token: token, email: user.email, id: user.id }
    });
  } catch (err) {
    /* istanbul ignore next */
    yield put({
      type: ActionTypes.USER_LOGIN_FAILURE,
      payload: err
    });
  }
}

/**
 * create
 */
export function* create(param) {
  const { email, password } = param.payload;
  try {
    const payload = { method: "POST", ...param };
    yield call(request, "/users", payload);

    yield put({
      type: ActionTypes.USER_LOGIN,
      payload: { email: email, password: password }
    });
  } catch (err) {
    /* istanbul ignore next */
    yield put({
      type: ActionTypes.USER_LOGIN,
      payload: { email: email, password: password }
    });
  }
}

/**
 * Logout
 */
export function* logout() {
  yield put({
    type: ActionTypes.USER_LOGOUT_SUCCESS
  });
}

/**
 * User Sagas
 */
export default function* root() {
  yield all([
    takeLatest(ActionTypes.USER_LOGIN, authenticate),
    takeLatest(ActionTypes.USER_LOGOUT, logout),
    takeLatest(ActionTypes.USER_CREATE, create)
  ]);
}
