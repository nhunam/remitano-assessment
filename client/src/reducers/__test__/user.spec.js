import reducer, { user } from "reducers";

import { ActionTypes } from "actions/user/type";

describe("User Reducers", () => {
  it("should return the initial state", () => {
    expect(reducer.user(undefined, {})).toMatchSnapshot();
  });

  it(`should handle ${ActionTypes.USER_LOGIN}`, () => {
    expect(
      reducer.user(undefined, { type: ActionTypes.USER_LOGIN })
    ).toMatchSnapshot();
  });

  it(`should handle ${ActionTypes.USER_CREATE}`, () => {
    expect(
      reducer.user(undefined, { type: ActionTypes.USER_CREATE })
    ).toMatchSnapshot();
  });
  it(`should handle ${ActionTypes.USER_LOGIN_SUCCESS}`, () => {
    const actions = {
      type: ActionTypes.USER_LOGIN_SUCCESS,
      payload: {
        data: {
          id: 1,
          name: "test",
          permissions: []
        }
      }
    };
    expect(reducer.user(undefined, actions)).toMatchSnapshot();
  });

  it(`should handle ${ActionTypes.USER_LOGOUT}`, () => {
    expect(
      reducer.user(undefined, { type: ActionTypes.USER_LOGOUT })
    ).toMatchSnapshot();
  });

  it(`should handle ${ActionTypes.USER_LOGOUT_SUCCESS}`, () => {
    expect(
      reducer.user(undefined, { type: ActionTypes.USER_LOGOUT_SUCCESS })
    ).toMatchSnapshot();
  });
});
