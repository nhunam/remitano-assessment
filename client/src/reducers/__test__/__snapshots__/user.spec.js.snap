// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Reducers should handle USER_CREATE 1`] = `
Object {
  "info": Object {
    "email": "",
    "id": -1,
    "token": "",
  },
  "isAuthenticated": false,
}
`;

exports[`User Reducers should handle USER_LOGIN 1`] = `
Object {
  "info": Object {
    "email": "",
    "id": -1,
    "token": "",
  },
  "isAuthenticated": false,
}
`;

exports[`User Reducers should handle USER_LOGIN_SUCCESS 1`] = `
Object {
  "info": Object {
    "email": undefined,
    "id": undefined,
    "token": undefined,
  },
  "isAuthenticated": true,
}
`;

exports[`User Reducers should handle USER_LOGOUT 1`] = `
Object {
  "info": Object {
    "email": "",
    "id": -1,
    "token": "",
  },
  "isAuthenticated": false,
  "status": undefined,
}
`;

exports[`User Reducers should handle USER_LOGOUT_SUCCESS 1`] = `
Object {
  "info": Object {
    "email": "",
    "id": -1,
    "token": "",
  },
  "isAuthenticated": false,
  "status": undefined,
}
`;

exports[`User Reducers should return the initial state 1`] = `
Object {
  "info": Object {
    "email": "",
    "id": -1,
    "token": "",
  },
  "isAuthenticated": false,
}
`;
